packages/dashboard/src/components/iot-dashboard/iot-context-menu/iot-dashboard-context-menu.tsx

```diff
@@ -0,0 +1,159 @@
+    }
+  }
+
+  disconnectedCallback() {
+    if (this.contextMenuTarget) {
+      this.contextMenuTarget.removeEventListener('contextmenu', this.onContextMenu.bind(this));
+      this.contextMenuTarget.removeEventListener('mousedown', this.onMouseDown.bind(this));
+    }
+  }
+
+  onContextMenu(event: MouseEvent) {
+    const { x, y } = getDashboardPosition(event);
+    this.contextMenuOpen = !this.contextMenuOpen;
+    this.contextMenuPosition = { x, y };
+    event.preventDefault();
+    return;
+  }
+
+  onMouseDown(event: MouseEvent) {
+    const isContextMenuGesture = event.button === MouseClick.Right;
+
+    if (!isContextMenuGesture && this.contextMenuOpen) {
+      this.hideContextMenu();
+    }
+  }
+
+  hideContextMenu() {
+    this.contextMenuOpen = false;
+    this.contextMenuPosition = undefined;
+  }
+
+  getContextMenuConfiguration = (): ContextMenuOptionConfiguration[][] => {
+    const withClose = (action: (event: MouseEvent) => void) => (event: MouseEvent) => {
+      action(event);
+      this.hideContextMenu();
+    };
+
+    const { copyText, pasteText, deleteText, bringToFrontText, sendToBackText } = Object.assign(
+      this.messageOverrides?.contextMenu ?? {},
+      DefaultDashboardMessages.contextMenu
+    );
+
+    const copyAction = withClose(() => this.copyWidgets());
+    const pasteAction = withClose(() => this.pasteWidgets(this.contextMenuPosition));
+    const deleteAction = withClose(() => this.deleteWidgets());
+    const bringToFrontAction = withClose(() => this.bringToFront());
+    const sendToBackAction = withClose(() => this.sendToBack());
+
+    const copyMenuOption = {
+      action: copyAction,
+      text: copyText,
+      hotkey: keyboardShortcuts.copy,
+      disabled: !this.hasSelectedWidgets,
+    };
+    const pasteMenuOption = {
+      action: pasteAction,
+      text: pasteText,
+      hotkey: keyboardShortcuts.paste,
+      disabled: !this.hasCopiedWidgets,
+    };
+    const deleteMenuOption = {
+      action: deleteAction,
+      text: deleteText,
+      disabled: !this.hasSelectedWidgets,
+    };
+    const bringToFrontMenuOption = {
+      action: bringToFrontAction,
+      text: bringToFrontText,
+      hotkey: keyboardShortcuts.bringToFront,
+      disabled: !this.hasSelectedWidgets,
+    };
+    const sendToBackMenuOption = {
+      action: sendToBackAction,
+      text: sendToBackText,
+      hotkey: keyboardShortcuts.sendToBack,
+      disabled: !this.hasSelectedWidgets,
+    };
+
+    return [
+      [copyMenuOption, pasteMenuOption, deleteMenuOption],
+      [bringToFrontMenuOption, sendToBackMenuOption],
+    ];
+  };
+
+  render() {
+    if (this.contextMenuOpen && this.contextMenuPosition) {
+      return (
+        <iot-context-menu {...this.contextMenuPosition}>
+          {this.getContextMenuConfiguration().map((section) => (
+            <iot-context-menu-section>
+              {section.map(({ text, hotkey, action, disabled }) => (
```

Comment:
I think we should add a key here, just in case `section` in the future could ever change based on other conditions. i.e. the list needs to update.

check out https://stenciljs.com/docs/templating-jsx section "Loops"

Reply:
Ya good catch, updated